---
title: "Gemcitabine/ATR inhibitor/combined treatments on DLBCL cell-lines"
author: Sherlyn Jemimah, PhD
output: html_notebook
---


### Initialize libraries and data
```{r}
## load libraries
library(DEGreport)
library(DESeq2)
library(ggplot2)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
library(tidyverse)

library(org.Hs.eg.db)
library(DOSE)
library(pathview)
library(clusterProfiler)
library(AnnotationHub)
library(ensembldb)
library(tidyverse)
library(annotables)

library(Cairo)

## load count data and sample information

#  files
count_file <- "D:/Allison_Jan2021/raw_counts.csv"
smple_info <- "D:/Allison_Jan2021/coldata.txt"

## count data
cts <- read.table(count_file,header=TRUE,sep=",",row.names=1)

# # if duplicate error
# rownames(cts) <- make.names(cts[,1],unique=TRUE)
# # if genes column is duplicated
# cts <- cts %>%
#   dplyr::select(-c(1))
# # if you get a DESeq non-integer error down the line
# cts <- round(cts,0) ## caution: some info may be lost due to this step.

## sample information/metadata
coldata <- read.table(smple_info,header=TRUE,sep="\t",row.names=1)
```


### Check count distribution
In general, a large set of genes will have very low (~0) expression.
```{r}
## Check all samples or pick randomly if there are too many samples
ggplot(cts) +
  geom_histogram(aes(x = HT01), stat = "bin", bins = 200) +
  xlim(-5, 500)  +
  xlab("Raw expression counts") +
  ylab("Number of genes")
```


## Checking dispersion
Plotting the mean against variance for the count data will provide an idea of which statistical model can be used.
Mean ~= variance implies Poisson model; mean < variance is best suited for Negative Binomial (NB) models.
```{r}
# input row numbers here (example, 1 an 2 are replicates of untreated HT cells)
replicate_indices <- c(2:3)

mean_counts <- apply(cts[, replicate_indices], 1, mean)
variance_counts <- apply(cts[, replicate_indices], 1, var)
df <- data.frame(mean_counts, variance_counts)

ggplot(df) +
        geom_point(aes(x=mean_counts, y=variance_counts)) + 
        geom_line(aes(x=mean_counts, y=mean_counts, color="red")) +
        scale_y_log10() +
        scale_x_log10()
```

Check that column names (samples) in count data exactly match the row names in the sample info file
```{r}
### Check that sample names match in both files
all(colnames(cts) %in% rownames(coldata))
all(colnames(cts) == rownames(coldata))
```


## Create DESeq object
```{r}
## Create DESeq2Dataset object
##immune genes
# immgenes <- as.vector(scan("D:/CRP-Proposal_Feb2021/nano_imgenes_all.txt",what=character(),sep="\n"))
# cts <- cts %>%
#   dplyr::filter(rownames(cts) %in% immgenes)
dds <- DESeqDataSetFromMatrix(countData=cts, colData=coldata, design = ~ Treatment)

head(counts(dds))
```


## Generate normalized counts
```{r}
# Check size factors 
dds <- estimateSizeFactors(dds)
sizeFactors(dds)

# Normalize counts
normalized_counts <- counts(dds, normalized=TRUE)
# save normalized data
write.table(normalized_counts, file="D:/Allison_Jan2021/normdata.txt", sep="\t", quote=F, col.names=NA)
```


## Quality Control: clustering analyses
The rlog() funtion can be a bit slow when you have e.g. > 20 samples. In these situations the vst() function is much faster and performs a similar transformation appropriate for use with plotPCA(). It's typically just a few seconds with vst() due to optimizations and the nature of the transformation.
```{r}
## Transform counts for data visualization
rld <- vst(dds, blind=TRUE)

## PCA plot

```

The plotPCA() returns values for PC1 and PC2 only. To check additional PCs, use the prcomp() function; see below.
```{r}
# Input is a matrix of log transformed values
rld <- varianceStabilizingTransformation(dds, blind=TRUE)
rld_mat <- assay(rld) ## assay() function from "SummarizedExperiment" package which loaded along with DESeq2
pca <- prcomp(t(rld_mat))

# Create data frame with metadata and PC1 and PC2 values for input to ggplot
df <- cbind(coldata, pca$x)
plot <- ggplot(df) + geom_point(aes(x=PC1, y=PC2, color=Full_State))
CairoPNG(file="D:/Allison_Jan2021/PCA-plot.png",
      units="in", 
      width=5, 
      height=4, 
      pointsize=12, 
      dpi=240)
print(plot)
dev.off()
```

For good quality data, the correlations between samples tend to be high (>0.99). Samples will also tend to cluster by group.
```{r}
# ### Compute pairwise correlation values
rld_cor <- cor(rld_mat)    ## cor() is a base R function
head(rld_cor)   ## check the output of cor(), make note of the rownames and colnames
# 
# ### Plot heatmap
pheatmap(rld_cor)
```

## Differential Expression analysis (DE)

```{r}
## Specify experiment design and reference
design <- ~ Treatment
#dds$Treatment <- relevel(dds$Treatment, ref="low")
## Create DESeq object
dds <- DESeqDataSetFromMatrix(countData=cts, colData=coldata, design=design)
## Run analysis
dds <- DESeq(dds)
```

```{r}
## Check the size factors (identical to estimateSizeFactors output?)
sizeFactors(dds)
## Total number of raw counts per sample
colSums(counts(dds))
## Total number of normalized counts per sample (depth of sequencing)
colSums(counts(dds, normalized=T))
```

## Gene-wise dispersion estimates
```{r}
## Plot dispersion estimates
plotDispEsts(dds)
```
## LFC shrinkage
```{r}
## Define contrasts, extract results table, and shrink the log2 fold changes
contrast_var <- c("Treatment", "Z", "C")
dds <- nbinomWaldTest(dds)
resultsNames(dds)

res     <- results(dds, contrast=contrast_var, alpha = 0.05)
reshrnk <- lfcShrink(dds,res=res,type="ashr")
```

Plot MA to compare the effects of LFC shrinkage.Check the results summary.
```{r}
# MA plots to check expression
plotMA(res, ylim=c(-2,2))
plotMA(reshrnk, ylim=c(-2,2))

# Results summary
summary(reshrnk)
```
Check the results table.
```{r}
reshrnk %>% data.frame() %>% na.omit() %>% dplyr::filter(padj.cutoff<0.05) %>% View()
```

## Get differentially expressed genes
```{r}
# Set thresholds
padj.cutoff <- 0.05
lfc.cutoff <- 2
# Tibble the results table
reshrnk_tb <- reshrnk %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
# subset to keep significant genes
sig <- reshrnk_tb %>%
        dplyr::filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
```

## Visualization single gene expression
```{r}
# Create tibbles for normalized counts and sample info, including row names
coldata1 <- coldata %>% 
  rownames_to_column(var="samplename") %>% 
  as_tibble()
normalized_counts1 <- normalized_counts %>% 
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()

## ggplot2 custom plot for single gene
# Save plotcounts to a data frame object
d <- plotCounts(dds, gene="MST1R", intgroup="Treatment", returnData=TRUE)

library(Cairo)

# Plotting the MOV10 normalized counts, using the samplenames (rownames of d as labels)
plot <- ggplot(d, aes(x = Treatment, y = count, color = Treatment)) + 
  geom_point(position=position_jitter(w = 0.1,h = 0)) +
  geom_text_repel(aes(label = rownames(d))) + 
  theme_bw() +
  ggtitle("MST1R") +
  theme(plot.title = element_text(hjust = 0.5))

CairoPNG(file="D:/Allison_Jan2021/MST1R.png",
      units="in", 
      width=5, 
      height=4, 
      pointsize=12, 
      dpi=240)
print(plot)
dev.off()

```
## Visualizing multiple genes
```{r}
# Order results by padj values
top20_genes <- reshrnk_tb %>%
        na.omit() %>%
        arrange(padj) %>% 	#Arrange rows by padj values
        pull(gene) %>% 		#Extract character vector of ordered genes
        head(n=30) 		#Extract the first 20 genes

# normalized counts for top 20 significant genes
top20_genes_norm <- normalized_counts1 %>%
        dplyr::filter(gene %in% c("CD1C","RGS8","MST1R","JAK3","NFKB2","CCR7","ITGAM","ARHGAP31","APOBEC3B","COL1A1","IL32"))

# gathering the columns to have normalized counts to a single column
top20_genes_gathered <- top20_genes_norm %>%
  gather(colnames(top20_genes_norm)[2:17], key = "samplename", value = "normalized_counts")

# check the column header in the "gathered" data frame
# View(top20_genes1)

# combine the sample info with gathered data
top20_genes_gathered <- inner_join(coldata1, top20_genes_gathered)

# plot using ggplot2
plot <- ggplot(top20_genes_gathered) +
        geom_point(aes(x = gene, y = normalized_counts, color = Treatment)) +
        scale_y_log10() +
        xlab("Genes") +
        ylab("log10 Normalized Counts") +
        ggtitle("Combined versus gemcitabine treatment") +
        theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(plot.title = element_text(hjust = 0.5))

CairoPNG(file="D:/Allison_Jan2021/ZG-vs-G.png",
      units="in", 
      width=8, 
      height=6, 
      pointsize=11, 
      dpi=300)
print(plot)

```
## Heatmap
```{r}
### Extract normalized expression for significant genes from the Z-treated and control samples, and set the gene column (1) to row names
norm_sig <- normalized_counts1 %>% 
              dplyr::filter(gene %in% sig$gene) %>% 
	      data.frame() %>%
	      column_to_rownames(var = "gene")

### Annotate our heatmap (optional)
annotation <- coldata1 %>% 
	dplyr::select(samplename, Full_State) %>% 
	data.frame(row.names = "samplename")

### Set a color palette
heat_colors <- brewer.pal(9, "YlGnBu")

### Run pheatmap
plot <- pheatmap(norm_sig,
                 color = heat_colors,
         cluster_rows = T, 
         cluster_cols = F,
         show_rownames = T,
         annotation = annotation, 
         border_color = NA, 
         fontsize = 10, 
         scale = "row", 
         fontsize_row = 10, 
         height = 20)

CairoPNG(file="D:/Allison_Jan2021/heatmap.png",
      units="in", 
      width=8, 
      height=8, 
      pointsize=11, 
      dpi=300)
print(plot)
```

## Volcano plot
```{r}
## Get logical vector where TRUE values denote padj values < 0.05 and fold change > 2 in either direction
restbl_ZvsC <- reshrnk_tb %>%
  mutate(threshold = padj < 0.05 & abs(log2FoldChange) >= 2)

## get which genes to label (top 20 by padj)
## Create a column to indicate which genes to label
restbl_ZvsC <- restbl_ZvsC %>% arrange(padj) %>% mutate(genelabels = "")
restbl_ZvsC$genelabels[1:nrow(restbl_ZvsC)] <- restbl_ZvsC$gene[1:nrow(restbl_ZvsC)]
restbl_ZvsC$genelabels[ -log10(restbl_ZvsC$padj) < 0.5] <- ""
restbl_ZvsC$genelabels[ abs(restbl_ZvsC$log2FoldChange) < 2] <- ""
View(restbl_ZvsC)

## Volcano plot
ggplot(restbl_ZvsC, aes(x = log2FoldChange, y = -log10(padj))) +
        geom_point(aes(colour = threshold)) +
        geom_text_repel(aes(label = genelabels)) +
        ggtitle("Changes caused by ATR inhibitor treatment") +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
```

## Pathway analysis
```{r}

library(msigdbr)

## immune genes
immgenes <- as.vector(scan("D:/CRP-Proposal_Feb2021/nano_imgenes_all.txt",what=character(),sep="\n"))

## Return the IDs for the gene symbols in the DE results
reshrnk_tb <- as.data.frame(reshrnk_tb)
# reshrnk_tb1 <- reshrnk_tb %>%
#   dplyr::filter(gene %in% immgenes)
rownames(reshrnk_tb) <- reshrnk_tb[,1]
# 
# idx <- grch37$symbol %in% rownames(reshrnk_tb)
# 
# ids <- grch37[idx, ]

## The gene names can map to more than one Ensembl ID (some genes change ID over time), 
## so we need to remove duplicate IDs prior to assessing enriched GO terms
# non_duplicates <- which(duplicated(ids$symbol) == FALSE)
# 
# ids <- ids[non_duplicates, ] 

## Merge the IDs with the results
h_gene_sets <- msigdbr::msigdbr(species = "Homo sapiens", category = "H")
res_ids <- inner_join(reshrnk_tb1, h_gene_sets, by=c("gene"="gene_symbol"))

## Create background dataset for hypergeometric testing using all genes tested for significance in the results                 
allOE_genes <- as.character(h_gene_sets$gene_symbol)

## Extract significant results
sigOE <- dplyr::filter(res_ids, padj < 0.05, abs(log2FoldChange) > 1.5)

sigOE_genes <- unique(as.character(sigOE$gene))

## Run GO enrichment analysis 
ego <- enrichGO(gene = sigOE_genes,
                universe = allOE_genes,
                keyType = "ENSEMBL",
                OrgDb = org.Hs.eg.db,
                ont = "BP",
                pAdjustMethod = "BH",
                qvalueCutoff = 0.05,
                readable = TRUE)
# 
# ego <- clusterProfiler::enricher(
#   gene=sigOE_genes,
#   pvalueCutoff = 0.05,
#   pAdjustMethod = "BH",
#   universe=h_gene_sets$gene_symbol,
#   minGSSize = 1,
#   maxGSSize = 500,
#   qvalueCutoff = 0.2,
#   TERM2GENE=h_gene_sets,
#   TERM2NAME=NA
# )
                
## Output results from GO analysis to a table
cluster_summary <- data.frame(ego)

# write.csv(cluster_summary, "D:/CRP-Proposal_Feb2021/clusterProfiler_rad51paths.csv")

## Dotplot 
plot <- clusterProfiler::dotplot(ego)

CairoPNG(file="D:/Allison_Jan2021/enrichment-analysis.png",
      units="in",
      width=12,
      height=12,
      pointsize=8,
      dpi=240)
print(plot)
dev.off()
# 
# 
# ## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms
# clusterProfiler::emapplot(ego, showCategory = 50)
```

Category netplot: Coloring the genes by log2 fold-changes
```{r}
## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector
sigOE <- sigOE %>%
  dplyr::filter(abs(log2FoldChange) > 2)
OE_foldchanges <- sigOE$log2FoldChange

names(OE_foldchanges) <- sigOE$gene

library(ggnewscale)
## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)
plot <- cnetplot(ego, 
         categorySize="pvalue", 
         showCategory = 5, 
         foldChange=OE_foldchanges, 
         vertex.label.font=6)

CairoPNG(file="D:/CRP-Proposal_Feb2021/RAD51-NETplot.png",
      units="in",
      width=18,
      height=10,
      pointsize=7,
      dpi=240)
print(plot)
dev.off()
         
## If some of the high fold changes are getting drowned out due to a large range, you could set a maximum fold change value
# OE_foldchanges <- ifelse(OE_foldchanges > 2, 2, OE_foldchanges)
# OE_foldchanges <- ifelse(OE_foldchanges < -2, -2, OE_foldchanges)
# 
# cnetplot(ego, 
#          categorySize="pvalue", 
#          showCategory = 5, 
#          foldChange=OE_foldchanges, 
#          vertex.label.font=6)
```
